app/src/openxr/cpp/OpenXRInputSource.cpp

```diff
@@ -356,11 +386,46 @@ XrResult OpenXRInputSource::SuggestBindings(SuggestedBindings& bindings) const
             assert(action != XR_NULL_HANDLE);
             RETURN_IF_XR_FAILED(CreateBinding(mapping.path, action, mSubactionPathName + "/" + axis.path, bindings));
         }
+
+        // Suggest binding for haptic feedback action.
+        RETURN_IF_XR_FAILED(CreateBinding(mapping.path, mHapticAction, mSubactionPathName + "/" + kPathHaptic, bindings));
     }
 
     return XR_SUCCESS;
 }
 
+void OpenXRInputSource::UpdateHaptics(ControllerDelegate &controller)
+{
+    uint64_t frameId = 0;
+    float pulseDuration = 0.0f;
+    float pulseIntensity = 0.0f;
+    controller.GetHapticFeedback(mIndex, frameId, pulseDuration, pulseIntensity);
+    if (frameId == 0 || pulseDuration <= 0.0f || pulseIntensity <= 0.0f) {
+        // No current haptic feedback, stop any ongoing haptic.
+        if (mStartHapticFrameId != 0) {
+            mStartHapticFrameId = 0;
+            VRB_LOG("stopHapticFeedback(%lu)", mStartHapticFrameId);
```

Comment:
I guess we don't need this debug code now.

Reply:
Right, removed.